### PR TITLE
Implement ammwrapper [approach 1]

### PR DESCRIPTION
**Pull request by doublespending**

### CODE REVIEW

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [inline contracts/AMMStrategy.sol:6]: `SafeMath` is not required in solidity ^0.8.0.

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [inline contracts/AMMStrategy.sol:4]: Remove

**Comment by charlesjhongc** [comment]
User `import { xxx } from "./XXX.sol` import pattern.

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [inline contracts/AMMStrategy.sol:10]: User tokenlon `Ownable` contract instead.

File: contracts/interfaces/IAMMStrategy.sol
Comment (charlesjhongc) [inline contracts/interfaces/IAMMStrategy.sol:3]: Remove

File: contracts/AMMStrategy.sol
Comment (pilagod) [reply contracts/AMMStrategy.sol:18]: If we go this way, then `setEntryPoint` could be removed together.

File: contracts/AMMStrategy.sol
Comment (pilagod) [inline contracts/AMMStrategy.sol:96]: I think maybe we could have a event for each operation, like cowswap

https://github.com/cowprotocol/contracts/blob/7efd5b43fbf46fe40bac9e6de1b6bab05d27c20b/src/contracts/GPv2Settlement.sol#L462-L466

This can log more info for each operation, and `opDests` could be removed in the swapped event.

File: test/forkMainnet/AMMStrategy/Setup.t.sol
Comment (pilagod) [inline test/forkMainnet/AMMStrategy/Setup.t.sol:147]:
```suggestion
        vm.startPrank(entryPoint);
```

File: test/forkMainnet/AMMStrategy/TradeWithMultiAMM.t.sol
Comment (pilagod) [inline test/forkMainnet/AMMStrategy/TradeWithMultiAMM.t.sol:27]:
```suggestion
        // entryPoint sends taker asset to ammStrategy before calling executeStrategy
```

File: test/forkMainnet/AMMStrategy/TradeWithMultiAMM.t.sol
Comment (pilagod) [inline test/forkMainnet/AMMStrategy/TradeWithMultiAMM.t.sol:46]: nit: Maybe we could use the test contract itself as the entry point, and then we can directly call strategy without pranking

**Review by pilagod — COMMENTED** [review]
nit: I think the tasks of this strategy are about (1) forward operation and calldata to external (2) ETH auto wrap/unwrap. So in my opinion the tests can be reduce to, for example, one multi AMM case (maybe 3 AMM uni, sushi, curve), two ETH as input/output cases, which are 3 cases in total.

Though testing each protocol separately can increase the confidence somehow, it would have more maintenance cost and doesn't focus enough on the core logic of this strategy.

Just my personal opinion, you can treat it as a minor reference. 😄

File: contracts/AMMStrategy.sol
Comment (doublespending) [inline contracts/AMMStrategy.sol:57]: * AMM cannot be an ERC-20 tokens, which user can change allowance of the tokens.
* For example, if WETH is an AMM, then user can `WETH.approve(SUSHISWAP,0)`

File: contracts/AMMStrategy.sol
Comment (doublespending) [inline contracts/AMMStrategy.sol:109]: * Wrap all native ETH first
* If operation need to use native ETH, native ETH will be withdraw in `_call` function

File: contracts/AMMStrategy.sol
Comment (doublespending) [inline contracts/AMMStrategy.sol:136]: Withdraw native ETH here if needed

File: contracts/AMMStrategy.sol
Comment (doublespending) [inline contracts/AMMStrategy.sol:74]: Uniswap universal router must use Permit2

File: contracts/AMMStrategy.sol
Comment (doublespending) [inline contracts/AMMStrategy.sol:171]: * unwrap WETH if outputToken is native ETH

File: contracts/AMMStrategy.sol
Comment (doublespending) [inline contracts/AMMStrategy.sol:85]: There may be some legacy tokens after swap.

File: contracts/AMMStrategy.sol
Comment (doublespending) [inline contracts/AMMStrategy.sol:151]: We should first approve permit2 and then approve in permit2

File: contracts/AMMStrategy.sol
Comment (pilagod) [reply contracts/AMMStrategy.sol:57]: Treat WETH as an AMM is a cool idea that I never thought before 👍

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [inline contracts/AMMStrategy.sol:13]: I think `IStrategy` can be removed here and @inheritdoc using `IAMMStrategy`.

File: contracts/AMMStrategy.sol
Comment (pilagod) [reply contracts/AMMStrategy.sol:151]: Q: So universal router will use plain `transferFrom` from permit2, without any permit signature?

**Review by pilagod — APPROVED** [review]

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [inline contracts/AMMStrategy.sol:107]: Since all of the balance will be transferred before return, there's no need to calculate before/after balance diff?

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [inline contracts/AMMStrategy.sol:124]: incorrect

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [inline contracts/AMMStrategy.sol:114]: Perhaps this check and event can be done within `_call()` function.

File: contracts/libraries/UniswapCommands.sol
Comment (charlesjhongc) [inline contracts/libraries/UniswapCommands.sol:6]: If this is only used in tests, move it under `test/libraries` instead.

File: test/forkMainnet/AMMStrategy.t.sol
Comment (charlesjhongc) [inline test/forkMainnet/AMMStrategy.t.sol:376]: User `setEOABalanceAndApprove()` and `deal()` instead. See `test/utils/BalanceUtil.sol`

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [reply contracts/AMMStrategy.sol:151]: Yes, universal router utilize allowance transfer instead of signature transfer.

Comment (doublespending) [reply contracts/AMMStrategy.sol:151]: Note: The UniversalRouter uses Permit2 to remove the need for token approvals being provided directly to the UniversalRouter. The Permit2 documentation can be found [here](https://docs.uniswap.org/contracts/permit2/overview).

I think so in command like `V2_SWAP_EXACT_IN`
* `address` The recipient of the output of the trade
* `uint256` The amount of input tokens for the trade
* `uint256` The minimum amount of output tokens the user wants
* `address[]` The UniswapV2 token path to trade along
* `bool` A flag for whether the input tokens should come from the msg.sender (through Permit2) or whether the funds are already in the UniversalRouter

File: contracts/AMMStrategy.sol
Comment (doublespending) [reply contracts/AMMStrategy.sol:107]: It is need just when `emit Swapped(inputToken, inputAmount, outputToken, balanceAfter - balanceBefore);`
If we don't emit the Swap event, It's ok.

File: contracts/AMMStrategy.sol
Comment (doublespending) [reply contracts/AMMStrategy.sol:13]: There will a compiler error: 
```
Compiler run failed
error[4682]: DocstringParsingError: Documentation tag @inheritdoc references contract "IAMMStrategy", but the contract does not contain a function that is overridden by this function.
  --> contracts/AMMStrategy.sol:97:5:
   |
97 |     /// @inheritdoc IAMMStrategy
```

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [reply contracts/AMMStrategy.sol:107]: The `balanceAfter - balanceBefore` here is equal to `erc20.balanceOf(address(this))`.

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [reply contracts/AMMStrategy.sol]: 8cff641

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [reply contracts/AMMStrategy.sol]: 8cff641

File: contracts/AMMStrategy.sol
Comment (charlesjhongc) [reply contracts/AMMStrategy.sol]: After discussion, remove `Swapped` event here and leave it to `GenericSwap`.

File: test/forkMainnet/AMMStrategy.t.sol
Comment (charlesjhongc) [reply test/forkMainnet/AMMStrategy.t.sol]: 6a2cdef

File: contracts/libraries/UniswapCommands.sol
Comment (charlesjhongc) [reply contracts/libraries/UniswapCommands.sol]: 6a2cdef

**Review by charlesjhongc — APPROVED** [review]

**Review by pilagod — DISMISSED** [review]

**Review by charlesjhongc — APPROVED** [review]

**Review by pilagod — APPROVED** [review]

**Pull request merged**
